src/inventory-service/inventory-core/src/main/java/com/inventory/core/InventoryItemService.java

```diff
@@ -71,95 +85,201 @@ public HandlerResponse<InventoryItemDTO> updateStock(UpdateInventoryStockRequest
+            }
+            return new HandlerResponse<>(null, List.of("Unknown error: " + e.getMessage()), false);
         }
     }
 
     public HandlerResponse<Boolean> reserveStockForOrder(String orderNumber, List<String> products, String conversationId) {
         final Span span = GlobalTracer.get().activeSpan();
+        logger.info("Reserving stock for order {} with {} products", orderNumber, products.size());
 
-        try {
-            var isFailure = false;
-            ArrayList<InventoryItem> stockAddedFor = new java.util.ArrayList<>();
-
+        if (span != null) {
             span.setTag("order.id", orderNumber);
             span.setTag("order.productCount", products.size());
+            span.setTag("order.conversationId", conversationId);
+        }
 
-            for (var productId : products) {
-                final Span stockCheckSpan = GlobalTracer.get()
-                        .buildSpan("stockCheck")
-                        .asChildOf(span)
-                        .start();
-
-                try (Scope scope = GlobalTracer.get().activateSpan(stockCheckSpan)) {
-                    stockCheckSpan.setTag("product.id", productId);
-                    var inventoryItem = this.repository.withProductId(productId);
+        // Store the products in the order cache early to ensure they're recorded even if reservation fails
+        try {
+            orderCache.store(orderNumber, new ArrayList<>(products));
+        } catch (Exception e) {
+            logger.error("Failed to store order in cache", e);
+            if (span != null) {
+                span.setTag("error.orderCache", true);
+                span.setTag("error.message", e.getMessage());
+            }
+            // Continue despite cache failure - we'll try to process the order anyway
+        }
 
-                    if (inventoryItem.getAvailableStockLevel() <= 0) {
-                        stockCheckSpan.setTag("product.outOfStock", "true");
-                        isFailure = true;
-                        break;
+        try {
+            AtomicBoolean isFailure = new AtomicBoolean(false);
+            List<InventoryItem> stockAddedFor = Collections.synchronizedList(new ArrayList<>());
+
+            // Process products in parallel for large orders for better performance
+            if (products.size() > 5) {
+                // For larger orders, process in parallel using CompletableFuture
+                logger.info("Using parallel processing for large order with {} products", products.size());
+                
+                List<CompletableFuture<Void>> futures = products.stream()
+                    .map(productId -> CompletableFuture.runAsync(() -> {
+                        if (isFailure.get()) {
+                            return; // Skip if we already know we'll fail
+                        }
+                        
+                        processProductReservation(productId, orderNumber, stockAddedFor, isFailure, span);
```

Comment:
Is this sync? If you wanted to go hard here, you'd want to backoff/retry/fail probably. Also, do you need compensating transactions in the case that a single product in the batch fails?

Reply:
It's handled by the fact that nothing persists to the database until after the failure check. So the reservation happens in memory, then it checks for failures, and then it persists if nothing fails. Good callout though, thanks man.